clang/lib/CodeGen/CGHLSLRuntime.cpp

```diff
@@ -536,89 +536,84 @@ void CGHLSLRuntime::generateGlobalCtorDtorCalls() {
   }
 }
 
-void CGHLSLRuntime::handleGlobalVarDefinition(const VarDecl *VD,
-                                              llvm::GlobalVariable *GV) {
-  // If the global variable has resource binding, add it to the list of globals
-  // that need resource binding initialization.
-  const HLSLResourceBindingAttr *RBA = VD->getAttr<HLSLResourceBindingAttr>();
-  if (!RBA)
-    return;
-
-  if (!HLSLAttributedResourceType::findHandleTypeOnResource(
-          VD->getType().getTypePtr()))
-    // FIXME: Only simple declarations of resources are supported for now.
-    // Arrays of resources or resources in user defined classes are
-    // not implemented yet.
-    return;
-
-  ResourcesToBind.emplace_back(VD, GV);
-}
-
-bool CGHLSLRuntime::needsResourceBindingInitFn() {
-  return !ResourcesToBind.empty();
+// Returns true if the type is an HLSL resource class
+static bool isResourceRecordType(const clang::Type *Ty) {
+  return HLSLAttributedResourceType::findHandleTypeOnResource(Ty) != nullptr;
 }
 
-llvm::Function *CGHLSLRuntime::createResourceBindingInitFn() {
-  // No resources to bind
-  assert(needsResourceBindingInitFn() && "no resources to bind");
-
+static void createResourceInitFn(CodeGenModule &CGM, const VarDecl *VD,
+                                 llvm::GlobalVariable *GV, unsigned Slot,
+                                 unsigned Space) {
   LLVMContext &Ctx = CGM.getLLVMContext();
   llvm::Type *Int1Ty = llvm::Type::getInt1Ty(Ctx);
 
-  llvm::Function *InitResBindingsFunc =
-      llvm::Function::Create(llvm::FunctionType::get(CGM.VoidTy, false),
-                             llvm::GlobalValue::InternalLinkage,
-                             "_init_resource_bindings", CGM.getModule());
+  llvm::Function *InitResFunc = llvm::Function::Create(
+      llvm::FunctionType::get(CGM.VoidTy, false),
+      llvm::GlobalValue::InternalLinkage,
+      ("_init_resource_" + VD->getName()).str(), CGM.getModule());
```

Comment:
I just noticed that in the existing code `_init_resource_bindings` is not marked as always inline, but I think it should. Can we add it here?

```suggestion
      ("_init_resource_" + VD->getName()).str(), CGM.getModule());
       InitResFunc->addFnAttr(llvm::Attribute::AlwaysInline);
```

Reply:
I think we can. It is eventually going to get inline anyway as do all internal function in HLSL, but it does not hurt to be specific here.

Reply:
If you expect that the function will be marked as always inline somewhere else, that does not happen. See this example: https://godbolt.org/z/5vfjExb71. There is still a call to `_init_resource_bindings`, and the function does not have the always inline attribute. Maybe there is something else that needs to be done then.

We don't have to do it this way, but we need to make sure it gets inlined at `-O0`.

Reply:
It is https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenHLSL/inline-functions.hlsl expect that I added an explicit register on the RWBuffer, which add the call to `_init_resource_bindings`. 

I came across this while working on something else.